corgidrp/recipe_templates/l1_to_l2a_nonlin.json

```diff
@@ -0,0 +1,34 @@
+{
+    "name" : "l1_to_l2a_nonlin",
+    "template" : true,
+    "drpconfig" : {
+        "track_individual_errors" : false
+    },
+    "inputs" : [],
+    "outputdir" : "",
+    "steps" : [
+        {
+            "name" : "prescan_biassub",
+            "keywords" : {
+                "return_full_frame" : true
+            }
+        },
+        {
+            "name" : "detect_cosmic_rays",
+            "calibs" : {
+                "DetectorParams" : "AUTOMATIC"
+            }
+        },
+        {
+            "name" : "calibrate_nonlin",
+            "keywords" : {
```

Comment:
They are the ones used to derive the TVAC calibration table (with a different s/w: Matlab). Default ones are the ones in the delivery from the engineering version in June, though we do not have the NL table that results from it with TVAC data. It was used with sims. Using these values instead of the default values, brings the peak-to-peak differences in the NL table coefficients from +/-3% to +/-2%.

Reply:
I have added the additional test_iit=true value to produce the smallest differences b/w the (Matlab) TVAC NL file and CORGI DRP's NL table (+/-1%). I'll use this for the assertion later on.
![CGI_EXCAM_L1_0000051731_NonLinearityCalibration](https://github.com/user-attachments/assets/890949b8-b03f-486e-88aa-48c9e5e7c514)

Reply:
We should be writing these recipes with the idea that this is what we'll be running on the real science data. So, I would prefer the default values be the ones we plan to use for the real on-sky data. If you need to change them, we should change them in the test (E.g., by passing in a custom, modified template).

Reply:
Agreed. That's what the template is doing. Scripts have default science values, not TVAC-comparison ones. I'll review it later today and confirm

Reply:
These are values used for the comparison with the TVAC NL table only. I'll change the template to be the default one for science, and create a specific one for the e2e test.

Reply:
Great. In that case, we probably want to have test that both recipes work.

Reply:
Differences between/w the delivered code in June for NL (engineering version) and CORGI DRP are plotted below for different values of DN and EM

![CGI_EXCAM_L1_0000051731_NonLinearityCalibration](https://github.com/user-attachments/assets/39e23dc1-8780-491a-8e46-1f545214252c)

Reply:
Differences of < 0.01% in nearly all cases seems pretty negligible to me. 

Do you have any thoughts to why there's still a bit of disagreement?

Reply:
I'm trying to change the ROI in one step to match the one used by the engineering code. I may contact Kevin bc he thought that these changes wouldn't matter. I reported these kind of differences back then (end of July). Maybe there's something else? I compared Guillermo's NL table (run in Windows/Intel) with mine (MacOS/M2) and relative differences are <= 4e-15, so that's double precision (I spotted a different behavior in numpy sort() b/w both operating systems though that raised some errors and ~1% differences. That's understood and fixed in the above plot)

I'll keep you posted

Reply:
I suggested some time ago to Marie that if there are (eventually) differences, say because bias subtraction is done at a different step in CORGI DRP (once and for all) compared to the engineering version, or else, we can always look at *the effect* of these differences in corrected data. if the differences when applying these tables are well below the error budget allowance for NL, then we'd be fine. Hopefully, we can reduce the differences too

Reply:
Quick update: using the same ROI in CORGI DRP as in the ENG version (one column difference) reduces the differences a lot except for the case of unity gain. I also checked that the mean frame is exactly the same on both codes. I'll investigate more on the gain=1 data.

![CGI_EXCAM_L1_0000051731_NonLinearityCalibration](https://github.com/user-attachments/assets/45b1db9e-d17c-487a-b135-57e4c0aa64cb)

Reply:
It's great to see that almost everything is agreeing. Kinda weird that EM=1 got worse... It seems like the first plot is actually better because everything is clearly in tolerance. 

You think the bias subtraction is done at a different step in corgidrp vs the engineering code? Can we move the prescan_biassub step in the recipe, or add a keyword toggle to prescan_biassub to match the engineering code?

What does ROI mean?

Reply:
ROI means region of interest. I wrote a summary of the comparison in Confluence: https://collaboration.ipac.caltech.edu/display/romancoronagraph/Documenting+Calibration+Function+Porting

Reply:
The decision made by DRP leads is to use the detector area from the engineering version of the code when deriving the NL correction table, instead of the one in detector_areas. This means to write the detector area in the NL recipe. I have committed the changes to comply with that, including cleaning up the e2e scripts that had other temporary options in consideration.

Reply:
The PR should be good to be merged. P.S. I'll resolve the conflicts on 9/12. Today I have another urgent task to work on